### PR TITLE
save_resume fix 2 (9)

### PR DESCRIPTION
**Pull request by trufanov-nok**
Few notes on this one (9): https://github.com/JohnLangford/vowpal_wabbit/issues/510

This patch affects GD and FTRL
I found 3 problems in GD if its state is restored from save_resume model:
a). average loss is different 
b). average loss since last is different in print_update() for first outputted value
c). number of examples per pass is wrong in vw' s summary

a) could be fixed by restoring `g.total_weight value`, b) - by restoring `sd->old_weighted_examples` and c). by restoring `all.current_pass` values.

as `g` object was removed from `save_load_online_state` arguments list (bcs of its call from ftrl, suppose)I've added `g*` as an optional argument. That required `save_load_online_state()` declaration change in `gd.h` and forward declaration of `struct gd;` before it. (As `struct gd` is declared and implemented solely in gd.cc. Moving it to .h caused some missing declarations for its fields)

I've defined new constant 
``#define VERSION_SAVE_RESUME_FIX "7.10.1"``
to compare while loading model. Current value in config.am is 7.8 which confused me a little, as latest ver should be 7.10. Perheps i'm doing something wrong here.

Warnings are shown only if both `--save_resume` and `-i` parameters are used and model file format is old. I'm checking save_resume field of file instead of related parameter of vw command line as it may be missing.

As model file version is checked in one .cc file and save_resume flag of model file in another .cc file I was forced to store model file version in all.model_file_version filed. This in its turn required a small modification in constructor of `version_struct``

### CODE REVIEW

**Comment by trufanov-nok** [comment]
After fix applied:

<pre>
truf@truf-laptop:~/ml/vowpal_wabbit$ vowpalwabbit/vw -k -c ./test/train-sets/rcv1_small.dat 
<b>--passes</b> 2 --save_resume --holdout_off                    
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = ./test/train-sets/rcv1_small.dat.cache
Reading datafile = ./test/train-sets/rcv1_small.dat
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
1.000000   1.000000          1      1.0    -1.0000   0.0000      128
0.893728   0.787455          2      2.0    -1.0000  -0.1126       44
0.905122   0.916517          4      4.0    -1.0000  -0.1701      190
0.924790   0.944458          8      8.0     1.0000  -0.0231       34
0.894742   0.864695         16     16.0     1.0000   0.0065       43
0.875196   0.855650         32     32.0    -1.0000   0.0423       47
0.838072   0.800947         64     64.0     1.0000   0.0619       54
0.754589   0.671106        128    128.0    -1.0000  -0.2779       67
0.662491   0.570393        256    256.0     1.0000   0.6543       86
0.564488   0.466484        512    512.0    -1.0000  -0.8828      104
<b>0.493744   0.423000</b>       1024   1024.0    -1.0000  -1.0000       58

finished run
<b>number of examples per pass = 1000</b>
passes used = 2
weighted example sum = 2000
weighted label sum = -164
<b>average loss = 0.285059</b>
best constant = -0.082
best constant's loss = 0.993276
total feature number = 157478


truf@truf-laptop:~/ml/vowpal_wabbit$ vowpalwabbit/vw -k -c ./test/train-sets/rcv1_small.dat 
<b>-f</b> /home/truf/temp/temp1.model --save_resume --holdout_off
final_regressor = /home/truf/temp/temp1.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
creating cache_file = ./test/train-sets/rcv1_small.dat.cache
Reading datafile = ./test/train-sets/rcv1_small.dat
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
1.000000   1.000000          1      1.0    -1.0000   0.0000      128
0.893728   0.787455          2      2.0    -1.0000  -0.1126       44
0.905122   0.916517          4      4.0    -1.0000  -0.1701      190
0.924790   0.944458          8      8.0     1.0000  -0.0231       34
0.894742   0.864695         16     16.0     1.0000   0.0065       43
0.875196   0.855650         32     32.0    -1.0000   0.0423       47
0.838072   0.800947         64     64.0     1.0000   0.0619       54
0.754589   0.671106        128    128.0    -1.0000  -0.2779       67
0.662491   0.570393        256    256.0     1.0000   0.6543       86
0.564488   0.466484        512    512.0    -1.0000  -0.8828      104

finished run
number of examples per pass = 1000
passes used = 1
weighted example sum = 1000
weighted label sum = -82
average loss = 0.502732
best constant = -0.082
best constant's loss = 0.993276
total feature number = 78739


truf@truf-laptop:~/ml/vowpal_wabbit$ vowpalwabbit/vw -k -c ./test/train-sets/rcv1_small.dat 
<b>-i</b> /home/truf/temp/temp1.model --save_resume --holdout_off
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
creating cache_file = ./test/train-sets/rcv1_small.dat.cache
Reading datafile = ./test/train-sets/rcv1_small.dat
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
<b>0.493744   0.423000</b>       1024   1024.0    -1.0000  -1.0000       58

finished run
<b>number of examples per pass = 1000</b>
passes used = 2
weighted example sum = 2000
weighted label sum = -164
<b>average loss = 0.285059</b>
best constant = -0.082
best constant's loss = 0.993276
total feature number = 157478

</pre>

**Comment by trufanov-nok** [comment]
Before fix is applied:
<pre>
truf@truf-laptop:~/ml/vowpal_wabbit$ vowpalwabbit/vw -k -c ./test/train-sets/rcv1_small.dat 
<b>-f</b> /home/truf/temp/temp1.model --save_resume --holdout_off
final_regressor = /home/truf/temp/temp1.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
creating cache_file = ./test/train-sets/rcv1_small.dat.cache
Reading datafile = ./test/train-sets/rcv1_small.dat
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
1.000000   1.000000          1      1.0    -1.0000   0.0000      128
0.893728   0.787455          2      2.0    -1.0000  -0.1126       44
0.905122   0.916517          4      4.0    -1.0000  -0.1701      190
0.924790   0.944458          8      8.0     1.0000  -0.0231       34
0.894742   0.864695         16     16.0     1.0000   0.0065       43
0.875196   0.855650         32     32.0    -1.0000   0.0423       47
0.838072   0.800947         64     64.0     1.0000   0.0619       54
0.754589   0.671106        128    128.0    -1.0000  -0.2779       67
0.662491   0.570393        256    256.0     1.0000   0.6543       86
0.564488   0.466484        512    512.0    -1.0000  -0.8828      104

finished run
number of examples per pass = 1000
passes used = 1
weighted example sum = 1000
weighted label sum = -82
average loss = 0.502732
best constant = -0.082
best constant's loss = 0.993276
total feature number = 78739


truf@truf-laptop:~/ml/vowpal_wabbit$ vowpalwabbit/vw -k -c ./test/train-sets/rcv1_small.dat <b>-i</b> /home/truf/temp/temp1.model --save_resume --holdout_off
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5

<b>WARNING: --save_resume functionality is known to have inaccuracy in model files version less than 7.10.1</b>

creating cache_file = ./test/train-sets/rcv1_small.dat.cache
Reading datafile = ./test/train-sets/rcv1_small.dat
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
<b>0.493931   0.211687</b>       1024   1024.0    -1.0000  -0.9849       58

finished run
<b>number of examples per pass = 2000</b>
passes used = 1
weighted example sum = 2000
weighted label sum = -164
<b>average loss = 0.286118<b>
best constant = -0.082
best constant's loss = 0.993276
total feature number = 157478

</pre>

**Comment by JohnLangford** [comment]
Merged in, thanks.
